pandas/io/common.py

```diff
@@ -391,13 +398,17 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
         elif encoding:
             # Python 3 and encoding
             f = open(path_or_buf, mode, encoding=encoding)
-        else:
+        elif is_text:
             # Python 3 and no explicit encoding
             f = open(path_or_buf, mode, errors='replace')
+        else:
+            # Python 3 and binary mode
+            f = open(path_or_buf, mode)
         handles.append(f)
 
     # in Python 3, convert BytesIO or fileobjects passed with an encoding
-    if compat.PY3 and (compression or isinstance(f, need_text_wrapping)):
+    if compat.PY3 and is_text and\
```

Comment:
`_get_handle` needs to deal with varies situations:

- py2 or py3
- binary (pickle, msgpack) or text (csv)
- if text, what's the encoding
- compression
- memory map
- open for read or write

maybe we can spilt `_get_handle` into two or more functions to make each single function simpler?